_posts/2023-01-12-January-Update-Finishing-my-Igalia-CE.md

```diff
@@ -0,0 +1,228 @@
+# IGT tests to V3D
+---
+
+Currently, V3D only has three basic IGT tests: `v3d_get_bo_offset`,
+`v3d_get_param`, and `v3d_mmap`. So, the basic goal of my CE project was to add
+more tests to the V3D driver.
+
+During the CE, I was able to develop tests for almost all eleven V3D ioctls
+(except `v3d_submit_tfu`). I began writing tests to the `v3d_create_bo` ioctl
+and *Performance Monitor* (perfmon) related ioctls. I developed tests that test
+the basic functionality of the ioctls and I inspected the kernel code to
+understand situations where the ioctl should fail.
+
+After those tests, I got the biggest challenge that I had on my CE project:
+performing a Mesa's noop job on IGT. The main problem I faced on submitting a
+noop job on IGT was: I would copy many and many Mesa files to IGT. And indeed,
+the final solution I came in with was to copy a couple of files from Mesa, but
+just three of them, which sounds reasonable. 
+
+So, after some time, I was able to bring the Mesa structure to IGT with minimal
+overhead. But, I was still not able to run a successful noop job as the job’s
+fence wasn't being signaled by the end of the job. Then, Melissa Wen had the
+great idea to run the CTS tests to inspect the noop job. With the CTS tests, I
+was able to hexdump the contents of the packet and understand what was going on
+wrong in my noop job.
+
+Running the CTS in the Raspberry Pi 4 was a fun side-quest of the project and
+ended up resulting in a commit to the CTS repository, as CTS wasn't handling
+appropriately the `wayland-scanner` for cross-compiling. CTS was picking the
+`wayland-scanner` from the host computer instead of picking the
+`wayland-scanner` executable available in the target sysroot. This was fixed
+with this simple patch:
+
+[**Allow override of wayland_scanner executable**](https://github.com/KhronosGroup/VK-GL-CTS/commit/6dfa0b69b46e69bdabd213ec2cf915bcd6e689f3)
+
+When I finally got a successful noop job, I was able to write the tests for the
+`v3d_submit_cl` and `v3d_wait_bo` ioctls. On these tests, I tested primarily job
+synchronization with single syncobjs and multiple syncobjs. In this part of the
+project, I had the opportunity to learn a lot about syncobjs and different forms
+of synchronization in the kernel and userspace.
+
+Having done the `v3d_submit_cl` tests, I developed the `v3d_submit_csd` tests in
+a similar way, as the job submission process is kind of similar. For submitting
+a CSD job, it is necessary to make a valid submission with a pipeline assembly
+shader and as IGT doesn't have a shader compiler, so I hard-coded the assembly
+of an empty shader in the code. In this way, I was able to get a simple CSD job
+submitted, and having done that, I could now play around with mixing CSD and CL
+jobs.
+
+In these tests, I could test the synchronization between two job queues and see,
+for example, if they were proceeding independently.
+
+So, by the end of the review process, we can expect V3D to have 72 IGT tests!
```

Comment:
V3D to have 72 IGT tests! -> expandir em: I'll add new XX tests, having in total 72 IGT tests!